packages/models/src/huggingface/text-generation.ts

```diff
@@ -0,0 +1,210 @@
+
+export namespace HfChatTypes {
+  // https://huggingface.co/docs/api-inference/detailed_parameters#text-generation-task
+  export type Request = {
+    model: string;
+    stream?: boolean;
+    inputs: string;
+    parameters?: {
+      top_k?: number;
+      top_p?: number;
+      temperature?: number;
+      repetition_penalty?: number;
+      max_new_tokens?: number;
+      // In seconds
+      max_time?: number;
+      return_full_text?: boolean;
+      num_return_sequences?: number;
+      do_sample?: boolean;
+    };
+    options?: {
+      use_cache?: boolean;
+      wait_for_model?: boolean;
+    };
+  };
+
+  export type RequestOptions = {
+    accessToken?: string;
+    apiUrl?: string;
+    fetch?: typeof fetch;
+  };
+
+  export type GeneratedText = {
+    generated_text: string;
+  };
+
+  // https://huggingface.co/docs/api-inference/detailed_parameters#text-generation-task
+  export type Response = GeneratedText | GeneratedText[];
+
+  // Best documentation I could find: https://huggingface.co/docs/huggingface_hub/main/en/package_reference/inference_client#huggingface_hub.inference._text_generation.TextGenerationStreamResponse
+  // I would like to find more formal documentation of their streaming API if we can.
+  export type Chunk = {
+    token: {
+      id: number;
+      text: string;
+      logprob: number;
+      special: boolean;
+    };
+    generated_text: string;
+    // Observed this but cannot find documentation
+    details: null;
+  };
+}
+
+async function run(
+  request: HfChatTypes.Request,
+  options: HfChatTypes.RequestOptions,
+): Promise<HfChatTypes.Response> {
+  const url = options.apiUrl || HF_MODEL_API_URL + request.model;
+
+  const headers_ = headers(options.accessToken);
+  const body = JSON.stringify({ ...request, stream: false });
+  const response = await POST(url, {
+    headers: headers_,
+    body,
+    fetch: options.fetch,
+  });
+
+  return response.json();
+}
+
+async function streamBytes(
+  request: HfChatTypes.Request,
+  options: HfChatTypes.RequestOptions,
+): Promise<ReadableStream<Uint8Array>> {
+  const url = options.apiUrl || HF_MODEL_API_URL + request.model;
+
+  const headers_ = headers(options.accessToken);
+  const body = JSON.stringify({ ...request, stream: true });
+  try {
+    const response = await POST(url, {
+      headers: headers_,
+      body,
+      fetch: options.fetch,
+    });
+
+    if (!response.body) {
+      throw new HttpError('Expected response body to be a ReadableStream', response);
+    }
+
+    return response.body;
+  } catch (e) {
+    if (e instanceof HttpError) {
+      const body = await e.response.json();
+      if (body?.error[0]?.includes('`stream` is not supported for this model')) {
+        throw new HttpError('This model does not support streaming', e.response);
+      }
+    }
+    throw e;
+  }
+}
+
+function noop(chunk: HfChatTypes.Chunk) {
+  return chunk;
+}
+
+// Extract the token text from a chunk. Spaces are part of the chunk, like:
```

Comment:
Nit: Spaces are part of the tokens for all the language models, so I don't think this comment is necessary.